### PR TITLE
Import Mujco mjcf files

### PR DESCRIPTION
**Pull request by Giulero**

### CODE REVIEW

**Comment by traversaro** [comment]
> Should I brake this?

Not an expert about this, but it is possible to leave `urdfstring` as a named argument and only use it if `model_string` is not set, perhaps printing a warning? A complext version of this in a form of a decorator is https://docs.astropy.org/en/latest/_modules/astropy/utils/decorators.html#deprecated_renamed_argument, but I guess we could just do something ad hoc. If that is too difficult, clearly explaining the change in the release notes and perhaps doing a bump in the minor release can help downstream users.

File: src/adam/pytorch/computations.py [hunk 1]
```diff
@@ -16,26 +19,70 @@ class KinDynComputations:
...
+
+    @staticmethod
+    def from_mujoco_xml(
+        xml_string: Union[str, Path],
+        joints_name_list: list = None,
+        gravity: torch.tensor = torch.tensor(
+            [0, 0, -9.80665, 0, 0, 0], dtype=torch.float64
+        ),
+    ) -> "KinDynComputations":
+        """Creates a KinDynComputations object from a Mujoco XML string
```
Comment (traversaro) [inline src/adam/pytorch/computations.py:72]: "Mujoco XML string" or "Mujoco XML path" or both?

File: src/adam/pytorch/computations.py [hunk 2]
```diff
@@ -16,26 +19,70 @@ class KinDynComputations:
...
+
+        Args:
+            xml_string (Union[str, Path]): The Mujoco XML path
```
Comment (traversaro) [inline src/adam/pytorch/computations.py:75]: path or string or both?

File: src/adam/pytorch/computations.py [hunk 3]
```diff
@@ -16,26 +19,70 @@ class KinDynComputations:
...
 
+    @staticmethod
+    def from_urdf(
+        urdf_string: Union[str, Path],
+        joints_name_list: list = None,
+        gravity: torch.tensor = torch.tensor(
+            [0, 0, -9.80665, 0, 0, 0], dtype=torch.float64
+        ),
+    ) -> "KinDynComputations":
+        """Creates a KinDynComputations object from a URDF string
```
Comment (traversaro) [inline src/adam/pytorch/computations.py:50]: string or path?

File: src/adam/pytorch/computations.py [hunk 4]
```diff
@@ -16,26 +19,70 @@ class KinDynComputations:
...
+        Args:
+            xml_string (Union[str, Path]): The Mujoco XML path
+            joints_name_list (list): list of the actuated joints
+            gravity (torch.tensor, optional): The gravity vector. Defaults to torch.tensor([0, 0, -9.80665, 0, 0, 0], dtype=torch.float64).
+
+        Returns:
+            KinDynComputations: The KinDynComputations object
+        """
+        return KinDynComputations(
+            model_string=xml_string, joints_name_list=joints_name_list, gravity=gravity
```
Comment (traversaro) [inline src/adam/pytorch/computations.py:83]: Is this actually supporting the mujoco xml string case?

File: tests/from_mujoco/test_Jax_mj.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
+# Copyright (C) 2021 Istituto Italiano di Tecnologia (IIT). All rights reserved.
+# This software may be modified and distributed under the terms of the
+# GNU Lesser General Public License v2.1 or any later version.
```
Comment (traversaro) [inline tests/from_mujoco/test_Jax_mj.py:3]: Wrong license (see https://github.com/ami-iit/adam/pull/64) and year. To be honest, I would remove the year from the copyright headers exactly to avoid this kind of errors.

**Comment by traversaro** [comment]
Minor: tests have a lot of commonalities, but I am not sure if there is a way to reduce the code duplication without hurting too much readability.

**Comment by traversaro** [comment]
General comment (not something that needs to be done): one way of being sure to support the full mjcf spec would be to change the parser to use the official mujoco parser, and actually create the adam model from the `mjModel` object populated by the official mujoco parser. Obviously the downside of this is that you add mujoco as a strict dependency if you want to use `mjcf` files.

**Comment by traversaro** [comment]
Nice to have: many users using mujoco use https://github.com/google-deepmind/mujoco_menagerie to get their models, if it is easy it may be useful to have a snippet that show in the README or the docs how to get a mujoco_menagerie model loaded in adam.

**Comment by S-Dafarra** [comment]
> > Should I brake this?
> 
> Not an expert about this, but it is possible to leave `urdfstring` as a named argument and only use it if `model_string` is not set, perhaps printing a warning? A complext version of this in a form of a decorator is [docs.astropy.org/en/latest/_modules/astropy/utils/decorators.html](https://docs.astropy.org/en/latest/_modules/astropy/utils/decorators.html#deprecated_renamed_argument), but I guess we could just do something ad hoc. If that is too difficult, clearly explaining the change in the release notes and perhaps doing a bump in the minor release can help downstream users.

This is also something that popped out with https://github.com/ami-iit/adam/pull/74

**Comment by Giulero** [comment]
Thanks @traversaro for the review, it really helped to clarify some doubts.

> The detection of the type of the file is based on extensions, but it can happen that URDF files as well use .xml as extension. However, this can be overridden (if I am not wrong) so it may not be a big problem. However, I would document this logic in the user facing function that the users actually call, i.e. KinDynComputations's init, to avoid misunderstandings. More complex logic could be use (like checking if the file contains a mujoco or robot top-level tag to distinguish between the two, but perhaps the existing logic may work fine until we find some exceptions.

and in particular
> like checking if the file contains a mujoco or robot top-level tag to distinguish between the two

I thought the same. My doubt was about the fact that the tags `mujoco` or `robot` are an unambiguous format and are always present in the descriptions. If so this is a better logic. I'm gonna implement it.

> Why is this supporting the use case of loading a urdf xml string, and not the a mjcf xml string? Probably as I user I would find confusing that one is supported and another not.

This is implementable as well!

> What is the failure mode if one tries to load an unsupported model? It raises an error or silently load a wrong file? I am particularly scared of the second case as it can lead to hard to detect bugs.

My idea was to raise an error if a unsupported field (e.g. `xyaxis`, `fullinertia`) is present in the description. 

> On a related note, perhaps we could add some helper functions to "normalize" mjcf files and/or string in adam itself? Perhaps those functions could lazy-load mujoco, to avoid adding a mujoco dependency to whole of adam, and only require mujoco if the conversion functions are used.

Indeed. Without adding the direct dependency, most likely, the user that is interested in this importer has already installed mujco as a dependency.

> General comment (not something that needs to be done): one way of being sure to support the full mjcf spec would be to change the parser to use the official mujoco parser, and actually create the adam model from the mjModel object populated by the official mujoco parser. Obviously the downside of this is that you add mujoco as a strict dependency if you want to use mjcf files.

Good idea! As in the previous comment, somehow the dependency can be handled supposing that the user needs to install mujoco anyway.

**Comment by JYChen18** [comment]
Hi,

I’m wondering if there’s any progress on this PR. 

The [mujoco_menagerie](https://github.com/google-deepmind/mujoco_menagerie) repo is a great source of high-quality robot models, so having support for them in adam would be super helpful.

**Comment by Giulero** [comment]
Hi @JYChen18, there is indeed progress! I just opened https://github.com/ami-iit/adam/pull/145 in favor of this PR. 
Once it is merged, there will be the possibility to load MuJoCo models.
